### PR TITLE
[torch][inductor] Extend KernelContextGuard profiling to the fallback and template kernel paths (#179699)

### PR DESCRIPTION
**Pull request by A-Kokolis**
Summary:

With `TORCHINDUCTOR_CPP_ENABLE_KERNEL_PROFILE=1`, AOTInductor emits a kernel
call inside a `{ ... }` block, and with `enable_kernel_context_guard` on top of
that the block opens with a `KernelContextGuard`. The guard publishes the
kernel's name and stack trace on a thread-local for the duration of the launch,
which is what lets a consumer attribute a launch back to the operator that
issued it. It is a different thing from the `RAIIAtenRecordFunctionHandle` that
gives a trace event its External id: a Triton kernel already gets one of those
in its generated launcher, and an extern kernel from the C shim path.

The guard reached only the kernels emitted through `codegen_node_schedule` and
`CppScheduling.flush`. Every other path that writes its own kernel call had
none, so its launches ran under whichever guard happened to still be live, or
under none.

 # Wrapper-level fallbacks

These bypass `generate_c_shim_extern_kernel_call` and emit their own call line:

- `_generate_scatter_fallback` (`aten::scatter_`, `scatter_reduce_`)
- `_generate_index_put_fallback` (`aten::index_put_`)
- `codegen_device_copy`, from both `DeviceCopy` and `InplaceCopyFallback`
(`aoti_torch_copy_`)

Nothing records these at all, so they get `profiled_kernel_scope`: the guard
plus a `RAIIAtenRecordFunctionHandle`, which is what the C shim path already
emits for an extern kernel. `profiled_kernel_scope` delegates the block to
`kernel_profile_scope` rather than opening it itself, and keeps its own gate --
`kernel_profile_enabled()`, which also requires a platform where the handle
exists -- passing it down so the block opens on exactly that condition.

 # Template and combo kernels

`codegen_template` writes its call line directly, and so does the combo-kernel
path:

- `SIMDScheduling.codegen_template`, both the single-kernel branch and the
`SizeHintMultiKernel` branch
- `SIMDScheduling.codegen_combo_kernel`
- `CppScheduling.codegen_template`
- `CUTLASSScheduling` and `ROCmCppScheduling`

Under max-autotune the template kernels are where most of the GPU time goes, so
this was the largest gap in guard coverage. Each site takes the same change the
non-template paths already use -- `kernel_profile_scope(kernel_name,
node_schedule)` around `call_kernel`, directly after the `codegen_comment` that
names it.

The guard rather than `profiled_kernel_scope`, because all of these already
carry a `RAIIAtenRecordFunctionHandle`: a Triton kernel in its generated
launcher and a C++ template from `CppTemplateKernel.maybe_codegen_profile`,
while CUTLASS and ROCm get one from the `triton=False` branch of
`CppWrapperGpu._generate_kernel_call_helper`. A second handle at the call site
would record them twice.

The remaining `codegen_template` implementations are deliberately left alone,
on one criterion: the backend cannot run under the C++ wrapper, so the scope
would be a no-op there. `FlyDSLScheduling` says so outright -- its template
returns `NotImplementedError("FlyDSL does not support AOT wrappers")` when
`cpp_wrapper` or `aot_mode` is set. `CuteDSLScheduling` and
`NVUniversalGemmScheduling` say it by construction: both call
`generate_kernel_call(triton=True)` with no `triton_meta`, which `CppWrapperGpu`
needs to build a launcher, and neither emits C++ for its kernel body. The same
holds for the Helion `ExternalTritonTemplateKernel` that `SIMDScheduling`'s
template branch can dispatch to, whose `call_kernel` writes Python.

For NVUniversalGemm the wrap would be worse than a no-op: its `call_kernel`
calls `codegen_allocation` for the `local_reduce` primary output, a real graph
buffer read after the call, so a block around it would scope that declaration
away the day the backend does support the C++ wrapper. Helion's is the same
shape and larger -- it codegens whole unfused epilogue nodes from inside
`call_kernel` -- but it sits on a branch that is already wrapped, inert for the
same reason.

 # Making the block a scope

The block is a real C++ scope, but the wrapper treated its braces as two
strings in a flat line list, so every memo that decides whether a declaration
is emitted reached across them. That produced generated code that does not
compile, in two ways:

- Buffer reuse. `memory_plan_reuse` resets its `MemoryPlanningState` only at a
subgraph boundary, so a workspace allocated and freed inside one block was
handed to the next block's allocation as `auto workspace_3 =
std::move(workspace_2);`, naming a variable the first block had destroyed.
Two persistent-TMA matmuls are enough: `error: use of undeclared identifier
'workspace_2'`.
- Lifetime. A `ZERO_PER_GRAPH` workspace -- the cooperative reduction
semaphores -- is allocated once at first use and never freed, so its
declaration landed inside whichever block asked for it first while every
later block named it: `error: use of undeclared identifier
'semaphores_cuda_0'`. This one reproduces without any of the wraps above.

`EnterKernelProfileScopeLine` / `ExitKernelProfileScopeLine` replace the two
strings and bracket the state the way `EnterSubgraphLine` / `ExitSubgraphLine`
already do for subgraphs: a fresh `MemoryPlanningState` in `memory_plan_reuse`,
and a snapshot of `computed_sizes` and `codegen_int_array_var_cache`. The two
memos are saved at different times on purpose -- `ensure_size_computed` runs
while lines are being collected, `codegen_int_array_var` while they are being
emitted -- so one is bracketed in `__post_init__` and the other in `codegen`.
The per-graph workspace cannot be fixed by scoping at all, so
`generate_workspace_allocation` emits it through
`writeline_outside_kernel_profile_scope`, which inserts in front of the block
instead of appending.

The cost is that a workspace no longer participates in buffer reuse while
profiling is on. Ordinary buffers are unaffected: the scheduler allocates them
before the block opens and frees them after it closes, so only the workspaces
`call_kernel` allocates internally are inside one.

`CppWrapperCpuArrayRef` plans its allocations with its own copy of
`memory_plan_reuse` rather than the base one, so it has to learn the two
bracket lines separately. Nothing reachable puts a memory-planning line between
those braces on a CPU ArrayRef compile today -- every CPU block wraps only
`call_kernel`, the scheduler emits frees after it, and every workspace
allocated inside a `call_kernel` is GPU-only -- so this is a divergence rather
than a live miscompile. It is fixed anyway because the copy rots silently: a
future CPU backend that allocates a workspace in `call_kernel`, or a block that
grows to contain a free, would reach it. Hoisting the shared body instead was
considered and rejected here: the base loop swaps `estimate_peak` per subgraph
and the ArrayRef one does not, and `codegen_while_loop` on that wrapper really
does emit subgraph lines, so unifying them would change generated code for
`while_loop` models under it -- a separate change wanting its own test.

 # Argument metadata

`write_record_function_handle` takes the same ordered
`profiling_args: list[str | None]` the C shim path uses -- one entry per
operator schema argument, a tensor handle or a placeholder. Both build the
conversions through `_profiling_ivalue_lines`, so the two cannot drift on the
order, the independent numbering of tensors and placeholders, or the variable
names; the shim path keeps its `out` entry by passing the output handle.
`scatter_reduce.two` records `[self, dim, index, src, reduce, include_self]`.

Three call sites reorder their arguments on the way to the kernel, so the
shared builder cannot derive the list for them, and they pass it themselves via
`make_profiling_args`: scatter keeps `dim` between its tensors, index_put's
index list is spread across the node's inputs and carries a None per unindexed
dimension, and `DeviceCopy`'s destination is the node's own output rather than
one of its inputs. `InplaceCopyFallback` is not one of them -- its inputs are
already `[dst, src]`, so the shared builder gets `copy_`'s schema order right
and it passes nothing. Each entry is derived separately from the IR node rather
than reusing the expression handed to the kernel: a reinterpret view's
expression owns the handle it names, so using it twice would free it once too
often.

`make_profiling_args` builds nothing unless the wrapper will record it, which is
what `records_profiling_args()` reports. That matters on
`CppWrapperCpuArrayRef`: an `ArrayRefTensor` is not an `AtenTensorHandle`, so
the ivalue conversion cannot be called on one, and deriving the list anyway
would make a reinterpret view mint a handle with no owner.

`scatter_fallback_kernel_name` gives the record and the emitted call one name:
`CppWrapperCpu` appends `_out` to the C shim symbol, the Python wrapper leaves
the resolved name alone. Trace consumers key on the emitted name, so recording
the ATen symbol instead would have kept scatter out of the analysis.

 # Not covered

- `generate_fallback_kernel_with_runtime_lookup`. Its three JIT variants
(`_nopython`, `_boxed`, `_python`) emit `RAIIAtenTensorHandle <output>;` at
the caller's scope and fill it inside their own block, so a profiling block
around the call would scope the fallback's outputs out of the code that
consumes them. The AOT proxy-executor variant carries no such declaration,
but the guard needs a kernel name and a node to take stack traces from, and
neither reaches `generate_fallback_kernel_with_runtime_lookup_aot`.
- The mix-order reduction call in `SIMDScheduling`, which passes
`deallocate_ws=False` precisely because the workspace outlives the call. The
scope bracket does not help there: the allocation and the free straddle the
brace rather than sitting inside it. It also has no kernel name to give a
guard -- it calls `codegen_comment` with `None`.
- `CppWrapperGpu.write_get_raw_stream` on the JIT cpp_wrapper path, where
`stream{idx}` is declared inline and memoized by an `lru_cache`, so a second
kernel in another block names a dead variable. AOT mode takes the stream as a
function parameter and dual-wrapper mode already hoists the declaration to the
entry prologue for exactly this reason, so only JIT `cpp_wrapper=True` on GPU
is exposed. The remedy is the same hoist, in the entry prologue rather than in
the memory planner, and it wants a JIT-cpp_wrapper regression test that this
file cannot host.
- `config.memory_planning`, off by default, plans through `MemoryPlanner`
instead of `memory_plan_reuse` and does not see the new bracket; a pool
handle first created inside a block would have the same problem.

 # Behaviour

Profiling off, generated code is unchanged: `kernel_profile_scope` writes
nothing when `cpp.enable_kernel_profile` is false and is a bare `yield` on the
Python wrapper, so neither bracket line is ever constructed, the hoist index
stays None and `writeline_outside_kernel_profile_scope` degenerates to the
`writeline` it replaced. `profiled_kernel_scope` reads its gate once and puts
both the block and the record behind it, and `make_profiling_args` returns
before it can emit codegen for a reinterpret view.

Test Plan:
Run from `fbcode/`.

```
buck2 test mode/opt fbcode//caffe2/test/inductor:test_aot_inductor -- \
--regex 'test_aoti_profiler|test_kernel_profile|KernelProfileNumelScope'
buck2 test mode/opt fbcode//caffe2/test/inductor:aot_inductor_arrayref_cpu_local -- \
--regex 'test_aoti_profiler|test_kernel_profile'
buck2 test mode/opt fbcode//caffe2/test/inductor:combo_kernels
```

62/0, 35/0 and 178/0 respectively.

`KernelProfileNumelScopeTest` now builds a real `CppWrapperCpu` over an empty
graph instead of hand-setting the attributes those paths happen to read, so a
path that starts reading a new one exercises it rather than raising
`AttributeError` and quietly testing nothing.

`KernelProfileScopeMemoryPlanningTest` covers the ArrayRef planning boundary.
It drives `CppWrapperCpuArrayRef.memory_plan_reuse` directly, with a probe
`MemoryPlanningLine` inside and outside a bracket pair, and asserts the two are
planned against different `MemoryPlanningState`s. An end-to-end test cannot
reach this: nothing puts a memory-planning line between those braces on a CPU
ArrayRef compile, so a model-level test patching
`aot_inductor.allow_stack_allocation` and `cpp.enable_kernel_profile` together
would pass with and without the change. Reverting the two branches makes this
test fail on its own assertion.

 # End to end

`mts_gpu_benchmark` then `aoti_trace_to_ir` on `1053126227:1024` (merge, H100,
kernel profiling on, which is the flow this stack exists to serve), compared
against the same flow run on the tree without these changes:

```
category      rows B/N     total_size B/N   size dlt
triton_tem    257/257      7004080746/7004080746   +0.00%
triton_poi     63/63       4298304070/4298304070   +0.00%
triton_red     89/89       2204982496/2204982496   +0.00%
extern_aten    35/35       1497586104/1497586104   +0.00%
GEMM (M,N,K) multiset fully conserved; identical dtype set
OVERALL: PASS
```

Two runs of the same tree also agree to +0.00% on every category, so the
comparison has no autotune noise to see through.

 # What the new tests pin, and how each was shown to fail

`test_kernel_profile_every_kernel_has_context` is the diff's claim stated
mechanically: for a graph mixing a max-autotune template, a pointwise kernel and
a reduction, every kernel the wrapper launches is named by a
`KernelContextGuard`. Launches are read off the `call_` launcher on GPU and the
`extern "C"` symbol on CPU, so a template's internal micro-gemm helpers -- which
the kernel body calls, not the wrapper -- are correctly out of scope. Reverting
the two `codegen_template` wraps fails it with the template named:
`['triton_tem_fused_add_mm_relu_t_0']` on cuda,
`['cpp_fused_add_mm_relu_t_0']` on cpu.

`test_kernel_profile_scope_is_a_scope_for_workspace_reuse` compiles two
persistent-TMA matmuls with profiling on. Reverting the scope bracket fails it
with `error: use of undeclared identifier 'workspace_2'`. It asserts a workspace
is actually present, so a future path that stops emitting one cannot make it
pass with nothing exercised, and that no `std::move` reuse names a source
declared in an earlier block -- reuse within one scope stays legal, so that,
rather than the presence of a reuse line, is the defect. It skips rather than
fails where the TMA template does not exist, since pinning the autotune choice
removes every other candidate.

`test_kernel_profile_scope_hoists_per_graph_workspace` compiles two cooperative
reductions. Reverting the bracket fails it with
`error: use of undeclared identifier 'semaphores_cuda_0'`. It then asserts the
semaphores declaration precedes the first guard, so a regression that emits it
inside a block is caught even if that build happens to succeed.

`test_kernel_profile_index_put_fallback` indexes one dimension, so the index
list carries a `None` and the record is
`[tensor, scalar, tensor, tensor, scalar]` -- an order the shared builder cannot
produce, since it would walk inputs then constants and drop the hole. Without
the wrap it fails with
`no profiled record function for aoti_torch_index_put_out`.

`test_kernel_profile_template_kernel` compiles a bias-free `nn.Linear` under
max-autotune with ATEN excluded, so the template kernel is the whole graph. It
asserts exactly one guarded kernel, since on CPU an ordinary fused kernel shares
the `cpp_` prefix and is guarded upstream. Without the wrap: `guarded: []`.

`KernelProfileNumelScopeTest.test_guard_blocks_nest_and_track_their_depth` now
codegens the collected lines and compares the emitted braces, rather than
comparing the line list against two literal strings.

 # Behaviour

Profiling off is unchanged by construction, not only by test: every new emission
sits inside `kernel_profile_scope`, which writes nothing when
`cpp.enable_kernel_profile` is false and is a bare `yield` on the Python
wrapper, so neither bracket line is constructed and the workspace hoist
degenerates to the append it replaced.

Profiling on, each test compiles and runs its model through `check_model` or
`AOTIRunnerUtil.compile`, so the generated C++ is shown to build and to produce
the right numbers with the extra block in place -- including the numel
redeclaration `kernel_profile_scope_depth` triggers inside a template kernel's
call, and the workspace reuse the bracket now suppresses.

 # Lint

The OSS `lintrunner` linters are not run by `arc lint` -- `arc f` reports "No
linters to run" for `fbcode/caffe2` -- so PYFMT and PYREFLY were run directly
against the changed files:

```
ruff format --config pyproject.toml --check <changed files>
ruff check  --config pyproject.toml        <changed files>
pyrefly check --config=pyrefly.toml        <changed .py>
```

`arc lint` is clean on all changed files, and each `fbcode/caffe2` file is
byte-identical to its `xplat/caffe2` mirror.

Differential Revision: D97207834

### CODE REVIEW

**Comment by desertfire** [comment]
@claude review this pr

**Comment by A-Kokolis** [comment]
> @claude review this pr

added the feedback from claude agent @desertfire

**Review by desertfire — APPROVED** [review]
